### PR TITLE
Reimplement #1652 - switchable time formats

### PR DESCRIPTION
**Pull request by daschuer**
This is a rebased Version of #1915 onto the original #1652 
to keep the original commit messages.

### CODE REVIEW

**Comment by daschuer** [comment]
LGTM, waiting for CI

File: src/util/duration.cpp
Comment (benis) [inline src/util/duration.cpp]: So should this not be using VERIFY_OR_DEBUG_ASSERT?

**Comment by benis** [comment]
There is still some work to do on this. Is there a way for me to push to this PR?

Also, the tests you said were failing - how do you run those?

**Comment by benis** [comment]
I would hope to have this done by the end of January for sure.

I've just created a PR to your hectoseconds branch with the change for Traditional (Coarse) time format.

**Comment by benis** [comment]
I still need to check if the tests are working and figure out why not if not. That's the only critical thing before this could be merged.

Here are some other things I would like to look at:
* Can we find a more conventional choice of separator character?

Also, the options displayed in the Deck Preferences are a bit inconsistent:
screenshot 2019-01-10 at 21 25 10
* is there a way to indicate that the hh is optional, or could we just remove it? Not many tracks over an hour long.
* 's zz' uses the defined centisecond separator character (thin space) - should it?
* 'k.sss zz' uses a static period, even though the actually-displayed character might be different. Then the defined centisecond separator character is used between sss and zz
* 'hs.ss zz' also has a static period but uses the hectosecond separator character between ss and zz
(See dlgprefdeck.cpp:122 onwards for relevant code.)

I also noticed that if you choose a different time format in Preferences but then hit Cancel, when you next load Preferences the drop-down has retained your value, so if you go back in to change something else, the deck preferences then get applied too.

**Comment by benis** [comment]
Just ran the tests, if I've understood the output correctly they now pass.

**Comment by daschuer** [comment]
Formating:
Let's follow strictly the SI recommendation in en-US format.
Decimal separator .
Thouthands colon , 
Decimal places seperator thin space.

Seconds ss,sss.ss
Kiloseconds ss.sss ss
Hectoseconds sss.ss ss

For the hh: I think a minute only format is also common. A movie runtime is given as 98 min for example. So we can pick IMHO that and drop the hh: entirely.

**Comment by benis** [comment]
I don't think your examples match up with SI. The [SI standard](https://en.wikipedia.org/wiki/ISO_31-0#Numbers) says that decimal places can be separated by a point or a comma (the former in US), and that the thousand separator should be a space. That makes sense to me, could we also use it as the hundred separator in the hectoseconds format?

**Comment by benis** [comment]
I've pushed the last of my commits. Here's how everything looks:

Preferences
prefs

Traditional
traditional

Traditional (Coarse)
traditional coarse

Seconds
seconds

Kiloseconds
kiloseconds

Hectoseconds
hectoseconds

File: src/preferences/dialog/dlgprefdeck.h
Comment (uklotzde) [inline src/preferences/dialog/dlgprefdeck.h]: Why do DisplayMode and DisplayFormat use inconsistent naming conventions for their enum literals?

File: src/util/duration.cpp
Comment (uklotzde) [inline src/util/duration.cpp]: Use a constant QString named kInvalidDurationString.

**Review by uklotzde — CHANGES_REQUESTED** [review]

**Comment by daschuer** [comment]
Traditional OK 
Traditional (Coarse) OK
Seconds OK 
Kiloseconds NOK  
* This is part of a kiloseconds movement https://kilosecond.info/ the 
* I think it should be 0.000 62  -1.000 95
hectoseconds I don't care much  
* I think it should be 0.00 62  -10.00 95 according to SI.

File: src/preferences/dialog/dlgprefdeck.h
Comment (benis) [reply src/preferences/dialog/dlgprefdeck.h]: Good question. You're referring to the fact that one is in Sentence Case and the other in SNAKE_CASE, right?

**Comment by benis** [comment]
> Traditional OK
> 
> Traditional (Coarse) OK
> 
> Seconds OK
> 
> Kiloseconds NOK
> 
>     * This is part of a kiloseconds movement https://kilosecond.info/ the
> 
>     * I think it should be 0.000 62  -1.000 95
> 
> 
> Hectoseconds I don't care much
> 
>     * I think it should be 0.00 62  -10.00 95 according to SI.

I still don't think that's consistent with SI, but feel free to change it.

File: src/preferences/dialog/dlgprefdeck.h
Comment (uklotzde) [reply src/preferences/dialog/dlgprefdeck.h]: Of course, it's obvious ;)

**Comment by benis** [comment]
https://en.wikipedia.org/wiki/International_System_of_Units#General_rules

* The symbol for the decimal marker is either a point or comma on the line
...
* Spaces should be used as a thousands separator (1 000 000) in contrast to commas or periods (1,000,000 or 1.000.000) to reduce confusion resulting from the variation between these forms in different countries.

So spaces for thousands, point or comma for decimal marker. You're proposing to do the opposite.

File: src/util/duration.h
Comment (benis) [inline src/util/duration.h]: I seem to have introduced this empty line in error.

**Comment by daschuer** [comment]
> So spaces for thousands, point or comma for decimal marker. You're proposing to do the opposite.

I read this like this:
1 ks = 16 min + 40 s
1 000 ks = 11 d + 13 h + 64 min + 40 s 
0.9 ks = 15 min 
0.001 ks = 1 s
0.000 001 hs = 1 ms 

1 hs = 1 min + 40 s
1 000 hs = 27 h + 46 min + 40 s 
0.9 hs = 1.5 min 
0.001 hs = 100 ms
0.000 01 hs = 1 ms

**Comment by benis** [comment]
Ah, I see where you're coming from now, sorry for the confusion. Are you happy to change it? It will be less faff than me submitting another PR to you.

**Comment by daschuer** [comment]
OK I will do the change, I am only unsure about hecto seconds. 
should we go with the SI correct:
0.000 1 hs = 10 ms
with the nicer looking: 
0.00 01 hs = 10 ms

**Comment by benis** [comment]
I vote for the second one. It's kind of an unconventional format so why not.

**Comment by daschuer** [comment]
OK, done. 

However, I have just played with it and it feels odd. 
I am considering to replace it with a plain seconds format with three leading zeros. 
000.00

Or should we reintroduce your original hectoseconds format? 
What do you think?

**Comment by benis** [comment]
I'm not attached to the hectoseconds format, so I wouldn't mind if we replaced it. I think your suggestion is a better choice and more likely to get used.

**Comment by daschuer** [comment]
OK I will change it.

**Comment by daschuer** [comment]
Done. What else is missing before merge?

**Comment by benis** [comment]
I think that's everything. I inadvertently introduced a blank line in src/util/duration.h (:17) so I guess this would be a good time to fix that, but otherwise this is good to go.

**Comment by daschuer** [comment]
OK, I am fine with this now. 
We may consider to alter also the library column accordingly, but that is a separate PR.

@uklotzde Is anything left to do before merge?

**Comment by daschuer** [comment]
Thank you @benis and @poelzi for working on this.

**Comment by uklotzde** [comment]
New bug: https://bugs.launchpad.net/mixxx/+bug/1830488

**Pull request merged**
